File: bindings/csharp/Serialization/JsonSerializer.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Reflection;

namespace MikanXR
{
	public class JsonSerializer
	{
		private class JsonWriteVisitor : IVisitor
		{
			private JToken _jsonToken;
			public JToken jsonToken => _jsonToken;

			public JsonWriteVisitor(JToken jsonToken)
			{
				_jsonToken = jsonToken;
			}

			public void visitClass(ValueAccessor accessor)
			{
				var objectValue = accessor.getValueObject();
				var objectType = accessor.ValueType;

				var jsonObject = new JObject();
				var jsonVisitor = new JsonWriteVisitor(jsonObject);
				Utils.visitObject(objectValue, objectType, jsonVisitor);

				setJsonValueFromJsonToken(accessor.ValueField, jsonObject);
			}

			public void visitList(ValueAccessor accessor)
			{
				// Get the list to be serialized
				var list = accessor.getValueObject() as IList;
				Type listType = accessor.ValueType;
				Type elementType = listType.GenericTypeArguments[0];

				// Serialize each element of the array
				JArray jsonArray = new JArray();
				foreach (var element in list)
				{
					// Make a fake "field" for an element in the array
					var elementAccessor = new ValueAccessor(element, elementType);

					// Serialize the element into a json object
					var elementVisitor = new JsonWriteVisitor(null);
					Utils.visitValue(elementAccessor, elementVisitor);

					// Add the element to the json array
					jsonArray.Add(elementVisitor.jsonToken);
				}

				setJsonValueFromJsonToken(accessor.ValueField, jsonArray);
			}

			public void visitDictionary(ValueAccessor accessor)
			{
				// Get the map to be serialized
				var map = accessor.getValueObject() as IDictionary;
				Type mapType = accessor.ValueType;
				Type keyType = mapType.GenericTypeArguments[0];
				Type valueType = mapType.GenericTypeArguments[1];

				// Serialize each key-value pair in the map
				JArray jsonPairArray = new JArray();
				IDictionaryEnumerator mapEnumerator = map.GetEnumerator();
				while (mapEnumerator.MoveNext())
				{
					// Serialize the key into a json object
					var keyObject = mapEnumerator.Key;
					var keyVisitor = new JsonWriteVisitor(null);
					var keyAccessor = new ValueAccessor(keyObject, keyType);
					Utils.visitValue(keyAccessor, keyVisitor);

					// Serialize the value into a json object
					var valueObject = mapEnumerator.Value;
					var valueVisitor = new JsonWriteVisitor(null);
					var valueAccessor = new ValueAccessor(valueObject, valueType);
					Utils.visitValue(valueAccessor, valueVisitor);

					var jsonPair = new JObject();
					jsonPair["key"]= keyVisitor.jsonToken;
					jsonPair["value"]= valueVisitor.jsonToken;

					// Add the pairs to the json array
					jsonPairArray.Add(jsonPair);
				}

				setJsonValueFromJsonToken(accessor.ValueField, jsonPairArray);
			}

			public void visitPolymorphicObject(ValueAccessor accessor)
			{
				// Get the SerializableObject container
				var serializableObject = accessor.getValueObject();
				var serializableObjectType = accessor.ValueType;

				// Get the instance and runtime class properties
				var instanceClassIdProperty = serializableObjectType.GetProperty("RuntimeClassId");
				var instanceProperty = serializableObjectType.GetProperty("Instance");

				var instanceClassId = (long)instanceClassIdProperty.GetValue(serializableObject);
				var instance = instanceProperty.GetValue(serializableObject);
				Type instanceType = instance.GetType();
				var instanceClassName = instanceType.Name;

				// Serialize the object into a json object
				var jsonRuntimeObject = new JObject();
				var jsonVisitor = new JsonWriteVisitor(jsonRuntimeObject);
				Utils.visitObject(instance, instanceType, jsonVisitor);

				var jsonSerializableObject = new JObject();
				jsonSerializableObject["class_name"] = instanceClassName;
				jsonSerializableObject["class_id"] = instanceClassId;
				jsonSerializableObject["value"] = jsonRuntimeObject;

				setJsonValueFromJsonToken(accessor.ValueField, jsonSerializableObject);
			}

			public void visitEnum(ValueAccessor accessor)
			{
				object enumObjectValue = accessor.getValueObject();
				string enumStringValue = Enum.GetName(accessor.ValueType, enumObjectValue);

				setJsonValueFromJsonToken(accessor.ValueField, enumStringValue);
			}

			public void visitBool(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<bool>(accessor);
			}

			public void visitByte(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<sbyte>(accessor);
			}

			public void visitUByte(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<byte>(accessor);
			}

			public void visitShort(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<short>(accessor);
			}

			public void visitUShort(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<ushort>(accessor);
			}

			public void visitInt(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<int>(accessor);
			}

			public void visitUInt(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<uint>(accessor);
			}

			public void visitLong(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<long>(accessor);
			}

			public void visitULong(ValueAccessor accessor)
			{
				throw new Exception(
					"JsonWriteVisitor::visitULong() " +
					"ULong Accessor " + accessor.ValueName +
					" type not supported by all JSON libraries");
			}

			public void visitFloat(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<float>(accessor);
			}

			public void visitDouble(ValueAccessor accessor)
			{
				setJsonValueFromAccessor<double>(accessor);
			}
			public void visitString(ValueAccessor accessor)
			{
				var sourceField = accessor.ValueField;
				string value = accessor.getValue<string>();

				setJsonValueFromJsonToken(sourceField, new JValue(value ?? string.Empty));
			}

			private void setJsonValueFromAccessor<T>(ValueAccessor accessor)
			{
				var sourceField= accessor.ValueField;
				T value = accessor.getValue<T>();

				setJsonValueFromJsonToken(sourceField, new JValue(value));
			}

			private void setJsonValueFromJsonToken(FieldInfo sourceField, JToken token)
			{
				if (sourceField != null)
				{
					var ownerJsonObject = _jsonToken as JObject;
					ownerJsonObject.Add(sourceField.Name, token);
				}
				else
				{
					_jsonToken = token;
				}
			}
		}

		public static string serializeToJsonString<T>(T instance) where T : class
		{
			return serializeToJsonString(instance, typeof(T));
		}

		public static string serializeToJsonString(object instance, Type structType)
		{
			JObject json = serializeToJson(instance, structType);
			return json.ToString();
		}

		public static JObject serializeToJson<T>(T instance) where T : class
		{
			return serializeToJson(instance, typeof(T));
		}

		public static JObject serializeToJson(object instance, Type structType)
		{
			var jsonObject = new JObject();
			var visitor = new JsonWriteVisitor(jsonObject);

			Utils.visitObject(instance, structType, visitor);

			return jsonObject;
		}
	}
}

File: bindings/typescript/Serialization/JsonSerializer.ts
import {
  ValueAccessor,
  IVisitor,
  visitObject,
  visitValue,
  FieldMetadata,
  getSerializationMetadata
} from './SerializationUtils.js';
import { PolymorphicObject } from '../PolymorphicObject.js';

/**
 * JSON write visitor for serializing objects to JSON
 */
class JsonWriteVisitor implements IVisitor {
  private jsonToken: any;

  constructor(jsonToken: any = {}) {
    this.jsonToken = jsonToken;
  }

  getJsonToken(): any {
    return this.jsonToken;
  }

  visitClass(accessor: ValueAccessor): void {
    const objectValue = accessor.getValueObject();
    const jsonObject: any = {};
    const jsonVisitor = new JsonWriteVisitor(jsonObject);

    visitObject(objectValue, accessor.valueType, jsonVisitor);

    this.setJsonValueFromJsonToken(accessor.valueField, jsonObject);
  }

  visitList(accessor: ValueAccessor): void {
    const list = accessor.getValueObject() as any[];
    const jsonArray: any[] = [];

    if (list) {
      for (const element of list) {
        // Check if element is a complex object with metadata
        const metadata = getSerializationMetadata(element);

        if (metadata && metadata.length > 0) {
          // Complex object - use visitor pattern
          const elementType = element.constructor;
          const elementVisitor = new JsonWriteVisitor({});
          visitObject(element, elementType, elementVisitor);
          jsonArray.push(elementVisitor.getJsonToken());
        } else {
          // Primitive or simple value - add directly
          jsonArray.push(element);
        }
      }
    }

    this.setJsonValueFromJsonToken(accessor.valueField, jsonArray);
  }

  visitDictionary(accessor: ValueAccessor): void {
    const record = accessor.getValueObject() as Record<any, any>;
    const jsonPairArray: any[] = [];

    if (record) {
      for (const key in record) {
        const value = record[key];
        let jsonKey = key;
        let jsonValue = value;

        // Serialize value if it's a complex object
        const valueMetadata = getSerializationMetadata(value);
        if (valueMetadata && valueMetadata.length > 0) {
          const valueVisitor = new JsonWriteVisitor({});
          visitObject(value, value.constructor, valueVisitor);
          jsonValue = valueVisitor.getJsonToken();
        }

        jsonPairArray.push({
          key: jsonKey,
          value: jsonValue
        });
      }
    }

    this.setJsonValueFromJsonToken(accessor.valueField, jsonPairArray);
  }

  visitPolymorphicObject(accessor: ValueAccessor): void {
    const polymorphicObject = accessor.getValueObject() as PolymorphicObject;
    const instance = polymorphicObject?.instance;
    const classId = polymorphicObject?.runtimeClassId || 0n;

    let className = '';
    const jsonRuntimeObject: any = {};

    if (classId !== 0n && instance) {
      className = instance.constructor.name;

      // Serialize the instance
      const jsonVisitor = new JsonWriteVisitor(jsonRuntimeObject);
      visitObject(instance, instance.constructor, jsonVisitor);
    }

    const jsonSerializableObject = {
      class_name: className,
      class_id: classId.toString(), // Convert bigint to string for JSON
      value: jsonRuntimeObject
    };

    this.setJsonValueFromJsonToken(accessor.valueField, jsonSerializableObject);
  }

  visitEnum(accessor: ValueAccessor): void {
    const enumValue = accessor.getValueObject();
    // Enums in TypeScript can be number or string
    this.setJsonValueFromJsonToken(accessor.valueField, enumValue);
  }

  visitBool(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<boolean>(accessor);
  }

  visitByte(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitUByte(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitShort(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitUShort(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitInt(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitUInt(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitLong(accessor: ValueAccessor): void {
    const value = accessor.getValue<bigint>();
    // Convert bigint to string for JSON compatibility
    this.setJsonValueFromJsonToken(accessor.valueField, value.toString());
  }

  visitULong(accessor: ValueAccessor): void {
    const value = accessor.getValue<bigint>();
    this.setJsonValueFromJsonToken(accessor.valueField, value.toString());
  }

  visitFloat(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitDouble(accessor: ValueAccessor): void {
    this.setJsonValueFromAccessor<number>(accessor);
  }

  visitString(accessor: ValueAccessor): void {
    const value = accessor.getValue<string>();
    this.setJsonValueFromJsonToken(accessor.valueField, value ?? '');
  }

  private setJsonValueFromAccessor<T>(accessor: ValueAccessor): void {
    const value = accessor.getValue<T>();
    this.setJsonValueFromJsonToken(accessor.valueField, value);
  }

  private setJsonValueFromJsonToken(sourceField: string | null, token: any): void {
    if (sourceField !== null) {
      this.jsonToken[sourceField] = token;
    } else {
      this.jsonToken = token;
    }
  }
}

/**
 * Serialize an object to JSON string
 */
export function serializeToJsonString(instance: any, instanceType?: any): string {
  const json = serializeToJson(instance, instanceType);
  return JSON.stringify(json);
}

/**
 * Serialize an object to JSON object
 */
export function serializeToJson(instance: any, instanceType?: any): any {
  const type = instanceType || instance?.constructor;
  const jsonObject: any = {};
  const visitor = new JsonWriteVisitor(jsonObject);

  visitObject(instance, type, visitor);

  return jsonObject;
}
